utils.go

```diff
@@ -90,19 +91,32 @@ var shutdownCallbacks []func() errCode
 // the general callback shutdowns
 var shutdownObjectStorageCallbacks []func() errCode
 
+// Protect shutdownCallbacks list from a concurrent access
+var shutdownMutex *sync.Mutex
+
 // Register callback functions that need to be called when process terminates.
 func registerShutdown(callback func() errCode) {
+	shutdownMutex.Lock()
 	shutdownCallbacks = append(shutdownCallbacks, callback)
+	shutdownMutex.Unlock()
 }
 
 // Register object storagecallback functions that need to be called when process terminates.
 func registerObjectStorageShutdown(callback func() errCode) {
+	shutdownMutex.Lock()
 	shutdownObjectStorageCallbacks = append(shutdownObjectStorageCallbacks, callback)
+	shutdownMutex.Unlock()
 }
 
 // Represents a type of an exit func which will be invoked during shutdown signal.
 type onExitFunc func(code int)
 
+// Initialize graceful shutdown mechanism
+func prepareGracefulShutdown() {
+	shutdownMutex = &sync.Mutex{}
```

Comment:
To have a mutex protection for 'callBack' list it is cleaner to put them in a struct together otherwise the mutex relation is harder to justify.